astropy/utils/data.py

```diff
@@ -1352,6 +1383,108 @@ def _release_download_cache_lock():
+    particularly those that 32-character hex strings or start with those,
+    probably indicates that something has gone wrong and inaccessible files
+    have accumulated in the cache. These can be removed with
+    `clear_download_cache`, either passking the filename returned here, or
+    with no arguments to  empty the entire cache and return it to a
+    reasonable, if empty, state.
+
+    Returns
+    -------
+    strays : set of strings
+        This is the set of files in the cache directory that do not correspond
+        to known URLs.
+    """
+    dldir, urlmapfn = _get_download_cache_locs()
+    # We're only reading but without the lock goodness knows what
+    # inconsistencies might be detected
+    _acquire_download_cache_lock()
+    try:
+        hash_files = set(os.path.join(dldir, k)
+                         for k in os.listdir(dldir))
+        try:
+            hash_files.remove(os.path.join(dldir, "urlmap"))
+        except KeyError:
+            pass
+        try:
+            hash_files.remove(os.path.join(dldir, "lock"))
+        except KeyError:
+            raise ValueError("Lock file missing!?")
+        with shelve.open(urlmapfn, flag="c") as url2hash:
+            for u, h in url2hash.items():
+                if not os.path.exists(h):
+                    msg = "URL '{}' points to nonexistent file '{}'".format(
+                            u, h)
+                    raise ValueError(msg)
+                try:
+                    hash_files.remove(h)
+                except KeyError:
+                    # Huh. Two different URLs retrieved identical files.
+                    pass
+                d, hexdigest = os.path.split(h)
+                if dldir != d:
+                    msg = "Expected downloaded files to be in '{}' but " +\
+                          "URL '{}' points to file '{}'".format(
+                                  dldir, u, h)
+                    raise ValueError(msg)
+                if check_hashes:
+                    hexdigest_file = compute_hash(h)
+                    if hexdigest_file != hexdigest:
+                        msg = "File corresponding to {} has contents that " +\
+                              "do not match the hash value: should be '{}' " +\
+                              "but is '{}'".format(
+                                      u, hexdigest, hexdigest_file)
+                        raise ValueError(msg)
+    finally:
+        _release_download_cache_lock()
+    return hash_files
+
+
+def _import_to_cache(url_key, filename,
+                     hexdigest=None,
+                     remove_original=False):
+    """Import the on-disk file specified by filename to the cache"""
+    # If this fails somehow might as well let the exception propagate
+    dldir, urlmapfn = _get_download_cache_locs()
+
+    if hexdigest is None:
+        hexdigest = compute_hash(filename)
+    _acquire_download_cache_lock()
```

Comment:
It seems like it would be nice to have the lock acquisition be a context manager...